File: python/julius/pricing.py
"""Offline, caller-supplied modeled USD pricing."""

from __future__ import annotations

from datetime import datetime, timezone
import math
from typing import Any


def _nonempty(value: Any) -> bool:
    return isinstance(value, str) and bool(value.strip())


def _count(value: Any) -> bool:
    return value is None or type(value) is int and 0 <= value <= 2**53 - 1


def _money(value: Any) -> bool:
    return value is None or type(value) in (int, float) and math.isfinite(value) and value >= 0


def _instant(value: Any) -> datetime | None:
    if not isinstance(value, str):
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None
    return parsed.astimezone(timezone.utc) if parsed.tzinfo is not None else None


def _snapshot(snapshot: dict[str, Any]) -> tuple[datetime, datetime | None]:
    if (
        not all(
            _nonempty(snapshot.get(key))
            for key in ("id", "modelId", "providerId", "source", "tier")
        )
        or snapshot.get("currency") != "USD"
    ):
        raise ValueError("Invalid price snapshot")
    effective = _instant(snapshot.get("effectiveAt"))
    expires = _instant(snapshot.get("expiresAt")) if "expiresAt" in snapshot else None
    if effective is None or ("expiresAt" in snapshot and (expires is None or expires <= effective)):
        raise ValueError("Invalid price snapshot")
    rates = snapshot.get("ratesPerMillion")
    if not isinstance(rates, dict) or any(
        key not in rates or not _money(rates[key])
        for key in ("inputUncached", "cacheRead", "cacheWrite", "output")
    ):
        raise ValueError("Invalid price rate")
    return effective, expires


def price_usage(usage: dict[str, Any], snapshot: dict[str, Any] | None) -> dict[str, Any]:
    call_id = usage.get("callId")
    if not _nonempty(call_id) or any(
        not _count(usage.get(key))
        for key in ("inputTokens", "cacheReadTokens", "cacheWriteTokens", "outputTokens")
    ):
        raise ValueError("Invalid usage for pricing")
    snapshot_id = usage.get("priceSnapshotId")
    if (
        not snapshot
        or not snapshot_id
        or snapshot.get("id") != snapshot_id
        or usage.get("modelId") != snapshot.get("modelId")
        or usage.get("providerId") != snapshot.get("providerId")
    ):
        return {"callId": call_id, "modeledCostUsd": None, "priceSnapshotId": snapshot_id}
    effective, expires = _snapshot(snapshot)
    occurred = _instant(usage.get("occurredAt"))
    if occurred is None or occurred < effective or (expires is not None and occurred >= expires):
        return {"callId": call_id, "modeledCostUsd": None, "priceSnapshotId": snapshot_id}
    input_tokens, read, write, output = (
        usage.get(key)
        for key in ("inputTokens", "cacheReadTokens", "cacheWriteTokens", "outputTokens")
    )
    if input_tokens is None or read is None or write is None or output is None:
        return {"callId": call_id, "modeledCostUsd": None, "priceSnapshotId": snapshot_id}
    uncached = input_tokens - read - write
    if uncached < 0:
        raise ValueError("Cache counters exceed input total")
    rates = snapshot["ratesPerMillion"]
    parts: tuple[tuple[int, float | int | None], ...] = (
        (uncached, rates["inputUncached"]),
        (read, rates["cacheRead"]),
        (write, rates["cacheWrite"]),
        (output, rates["output"]),
    )
    if any(tokens > 0 and rate is None for tokens, rate in parts):
        return {"callId": call_id, "modeledCostUsd": None, "priceSnapshotId": snapshot_id}
    cost = sum(tokens * (rate or 0) / 1_000_000 for tokens, rate in parts)
    if not math.isfinite(cost):
        raise ValueError("Modeled cost overflow")
    return {"callId": call_id, "modeledCostUsd": cost, "priceSnapshotId": snapshot_id}


def compare_modeled_costs(
    baseline: dict[str, Any],
    usage: list[dict[str, Any]],
    overhead: list[dict[str, Any]] | None = None,
    coverage_complete: bool = False,
) -> dict[str, Any]:
    overhead = [] if overhead is None else overhead
    if (
        not _nonempty(baseline.get("id"))
        or not _nonempty(baseline.get("comparisonScope"))
        or baseline.get("evidence") not in ("reconstructed_baseline", "controlled_experiment")
        or not _money(baseline.get("modeledCostUsd"))
    ):
        raise ValueError("Invalid comparable baseline")
    call_ids: set[str] = set()
    for item in usage:
        call_id = item.get("callId")
        if not _nonempty(call_id) or call_id in call_ids or not _money(item.get("modeledCostUsd")):
            raise ValueError("Invalid or duplicate call ID")
        call_ids.add(str(call_id))
    overhead_ids: set[str] = set()
    for item in overhead:
        item_id = item.get("id")
        if (
            not _nonempty(item_id)
            or item_id in overhead_ids
            or item.get("kind") not in ("local", "external")
            or not _money(item.get("costUsd"))
        ):
            raise ValueError("Invalid or duplicate overhead ID")
        overhead_ids.add(str(item_id))
    observed = (
        sum(item["modeledCostUsd"] for item in usage)
        if (usage or coverage_complete)
        and all(item.get("modeledCostUsd") is not None for item in usage)
        else None
    )
    overhead_cost = (
        sum(item["costUsd"] for item in overhead)
        if all(item.get("costUsd") is not None for item in overhead)
        else None
    )
    if any(value is not None and not math.isfinite(value) for value in (observed, overhead_cost)):
        raise ValueError("Modeled cost overflow")
    base_cost = baseline.get("modeledCostUsd")
    net = (
        base_cost - observed - overhead_cost
        if base_cost is not None and observed is not None and overhead_cost is not None
        else None
    )
    if net is not None and not math.isfinite(net):
        raise ValueError("Modeled cost overflow")
    return {
        "baselineId": baseline["id"],
        "baselineEvidence": baseline["evidence"],
        "baselineModeledCostUsd": base_cost,
        "observedModeledCostUsd": observed,
        "overheadUsd": overhead_cost,
        "netModeledSavingsUsd": net,
    }

File: src/pricing.ts
import type { Evidence } from "./events";

export interface PriceSnapshot {
  readonly id: string;
  readonly modelId: string;
  readonly providerId: string;
  readonly source: string;
  readonly effectiveAt: string;
  readonly currency: "USD";
  readonly tier: string;
  /** Caller supplied USD per million tokens. Null means unavailable. */
  readonly ratesPerMillion: Readonly<{
    inputUncached: number | null;
    cacheRead: number | null;
    cacheWrite: number | null;
    output: number | null;
  }>;
}

export interface PricedUsage {
  readonly callId: string;
  readonly modelId: string | null;
  readonly providerId: string | null;
  readonly inputTokens: number | null;
  readonly cacheReadTokens: number | null;
  readonly cacheWriteTokens: number | null;
  readonly outputTokens: number | null;
  readonly priceSnapshotId: string | null;
}

export interface ModeledCost {
  readonly callId: string;
  readonly modeledCostUsd: number | null;
  readonly priceSnapshotId: string | null;
}

export interface ExplicitOverhead {
  readonly id: string;
  readonly kind: "local" | "external";
  readonly costUsd: number | null;
}

export interface ComparableBaseline {
  readonly id: string;
  readonly evidence: Extract<Evidence, "reconstructed_baseline" | "controlled_experiment">;
  readonly modeledCostUsd: number | null;
  readonly comparisonScope: string;
}

const validCount = (value: unknown): boolean => value === null || typeof value === "number" && Number.isSafeInteger(value) && value >= 0;
const validMoney = (value: unknown): boolean => value === null || typeof value === "number" && Number.isFinite(value) && value >= 0;
const nonempty = (value: string): boolean => value.trim().length > 0;

function validateSnapshot(snapshot: PriceSnapshot): void {
  if (![snapshot.id, snapshot.modelId, snapshot.providerId, snapshot.source, snapshot.tier].every(nonempty) ||
    snapshot.currency !== "USD" || Number.isNaN(Date.parse(snapshot.effectiveAt))) throw new Error("Invalid price snapshot");
  if (!snapshot.ratesPerMillion || !["inputUncached", "cacheRead", "cacheWrite", "output"].every(key =>
    Object.hasOwn(snapshot.ratesPerMillion, key) && validMoney(snapshot.ratesPerMillion[key as keyof PriceSnapshot["ratesPerMillion"]]))) throw new Error("Invalid price rate");
}

/** Computes a modeled token cost, never a provider measurement. */
export function priceUsage(usage: PricedUsage, snapshot: PriceSnapshot | null): ModeledCost {
  if (!nonempty(usage.callId) || ![usage.inputTokens, usage.cacheReadTokens, usage.cacheWriteTokens, usage.outputTokens].every(validCount)) throw new Error("Invalid usage for pricing");
  if (!snapshot || !usage.priceSnapshotId || snapshot.id !== usage.priceSnapshotId ||
    usage.modelId !== snapshot.modelId || usage.providerId !== snapshot.providerId) {
    return { callId: usage.callId, modeledCostUsd: null, priceSnapshotId: usage.priceSnapshotId };
  }
  validateSnapshot(snapshot);
  const { inputTokens, cacheReadTokens, cacheWriteTokens, outputTokens } = usage;
  if (inputTokens === null || cacheReadTokens === null || cacheWriteTokens === null || outputTokens === null) {
    return { callId: usage.callId, modeledCostUsd: null, priceSnapshotId: snapshot.id };
  }
  const uncached = inputTokens - cacheReadTokens - cacheWriteTokens;
  if (uncached < 0) throw new Error("Cache counters exceed input total");
  const items: readonly [number, number | null][] = [
    [uncached, snapshot.ratesPerMillion.inputUncached],
    [cacheReadTokens, snapshot.ratesPerMillion.cacheRead],
    [cacheWriteTokens, snapshot.ratesPerMillion.cacheWrite],
    [outputTokens, snapshot.ratesPerMillion.output],
  ];
  if (items.some(([tokens, rate]) => tokens > 0 && rate === null)) return { callId: usage.callId, modeledCostUsd: null, priceSnapshotId: snapshot.id };
  const cost = items.reduce((sum, [tokens, rate]) => sum + tokens * (rate ?? 0) / 1_000_000, 0);
  if (!Number.isFinite(cost)) throw new Error("Modeled cost overflow");
  return { callId: usage.callId, modeledCostUsd: cost, priceSnapshotId: snapshot.id };
}

export interface Comparison {
  readonly baselineId: string;
  readonly baselineEvidence: ComparableBaseline["evidence"];
  readonly baselineModeledCostUsd: number | null;
  readonly observedModeledCostUsd: number | null;
  readonly overheadUsd: number | null;
  readonly netModeledSavingsUsd: number | null;
}

/** Signed modeled comparison. Usage and overhead IDs must be unique within the supplied scope. */
export function compareModeledCosts(baseline: ComparableBaseline, usage: readonly ModeledCost[], overhead: readonly ExplicitOverhead[] = [], coverageComplete = false): Comparison {
  if (!nonempty(baseline.id) || !nonempty(baseline.comparisonScope) ||
    !["reconstructed_baseline", "controlled_experiment"].includes(baseline.evidence) ||
    !validMoney(baseline.modeledCostUsd)) throw new Error("Invalid comparable baseline");
  const callIds = new Set<string>(), overheadIds = new Set<string>();
  for (const item of usage) {
    if (!nonempty(item.callId) || callIds.has(item.callId) || !validMoney(item.modeledCostUsd)) throw new Error("Invalid or duplicate call ID");
    callIds.add(item.callId);
  }
  for (const item of overhead) {
    if (!nonempty(item.id) || overheadIds.has(item.id) || !["local", "external"].includes(item.kind) || !validMoney(item.costUsd)) throw new Error("Invalid or duplicate overhead ID");
    overheadIds.add(item.id);
  }
  const observed = (usage.length > 0 || coverageComplete) && usage.every(item => item.modeledCostUsd !== null)
    ? usage.reduce((sum, item) => sum + (item.modeledCostUsd ?? 0), 0) : null;
  const overheadCost = overhead.every(item => item.costUsd !== null)
    ? overhead.reduce((sum, item) => sum + (item.costUsd ?? 0), 0) : null;
  if ((observed !== null && !Number.isFinite(observed)) || (overheadCost !== null && !Number.isFinite(overheadCost))) throw new Error("Modeled cost overflow");
  const net = baseline.modeledCostUsd !== null && observed !== null && overheadCost !== null
    ? baseline.modeledCostUsd - observed - overheadCost : null;
  if (net !== null && !Number.isFinite(net)) throw new Error("Modeled cost overflow");
  return { baselineId: baseline.id, baselineEvidence: baseline.evidence,
    baselineModeledCostUsd: baseline.modeledCostUsd, observedModeledCostUsd: observed,
    overheadUsd: overheadCost, netModeledSavingsUsd: net };
}
